### ISSUE TITLE
pdf.Text widget with justify dosen't left align last line of paragraphs except last one.

### ISSUE THREAD

**Issue by jeanadam** [DavBfr/dart_pdf#281]
**Describe the bug**
When using pdf.Text widget with textAlign = textAlign.justify to place a justified text on PDF, if the text has multiple paragraphs, like "Phrase 1 ... end paragraph 1.**\n\n**Phrase 2... end paragraph 2.". The last line of paragraph 1 is not left align, as it happens with material text widget when displaying same text and using justify. The code only leftAlign the last line of the whole text, not considering paragraphs in the middle.

I did some debugging and figured there is a bug with the pdf\lib\text.dart line 487, when the paragraph's last line has only 1 word, the behavior changes because spans.length is 1 and delta becames **infinity** [variable / (spans.length - 1)]. So when the last line of the paragraph has just 1 word, it will left align. It is an unstabble situation.

I already think of a fix, checking which lines have '\n\n' and make it last = true when current line +1 is in this array of lines.

Code where part of the problem happens (pdf\lib\text.dart):
```dart
 case TextAlign.justify:
        if (last) {
          totalWidth = wordsWidth;
          break;
        }
        delta = (totalWidth - wordsWidth) / (spans.length - 1);
        double x = 0;
        for (_Span span in spans) {
          span.offset = span.offset.translate(x, -baseline);
          x += delta;
        }
        return totalWidth;
```

Code where fix need to be done (pdf\lib\text.dart):
```dart
 for (int line = 0; line < spanLines.length; line++) {
          for (String word in spanLines[line].split(RegExp(r'\s'))) {
...
  }
}
```

**To Reproduce**
Code snippet to reproduce the behavior (main.dart):
```dart
import 'package:flutter/material.dart';

import 'dart:io';

import 'package:pdf/pdf.dart';
import 'package:pdf/widgets.dart' as pw;
import 'package:path_provider/path_provider.dart';

import 'package:open_file/open_file.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        // This is the theme of your application.
        //
        // Try running your application with "flutter run". You'll see the
        // application has a blue toolbar. Then, without quitting the app, try
        // changing the primarySwatch below to Colors.green and then invoke
        // "hot reload" (press "r" in the console where you ran "flutter run",
        // or simply save your changes to "hot reload" in a Flutter IDE).
        // Notice that the counter didn't reset back to zero; the application
        // is not restarted.
        primarySwatch: Colors.blue,
      ),
      home: MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key, this.title}) : super(key: key);

  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  int _counter = 0;

  Future<String> get _localPath async {
    final directory = await getApplicationDocumentsDirectory();
    return directory.path;
  }

  void _incrementCounter() async {
    setState(() {
      _counter++;
    });

    final pw.Document doc = pw.Document();
    doc.addPage(pw.MultiPage(
        pageFormat:
        PdfPageFormat.a5,
        crossAxisAlignment: pw.CrossAxisAlignment.start,
        build: (pw.Context context) => <pw.Widget>[
    pw.Header(
    level: 0,
        child: pw.Row(
            mainAxisAlignment: pw.MainAxisAlignment.spaceBetween,
            children: <pw.Widget>[
              pw.Text('Portable Document Format', textScaleFactor: 2),
              pw.PdfLogo()
            ])),
        pw.Paragraph(
            text:
            'The Portable Document Format (PDF) is a file format developed by Adobe in the 1990s to present documents, including text formatting and images, in a manner independent of application software, hardware, and operating systems. Based on the PostScript language, each PDF file encapsulates a complete description of a fixed-layout flat document, including the text, fonts, vector graphics, raster images and other information needed to display it. PDF was standardized as an open format, ISO 32000, in 2008, and no longer requires any royalties for its implementation.',),
          pw.Text('Aa aaaaaaa aaaaa a aaaaa aa aaaaaaaa aaaaaa aa aaaaa aaa aaaaaa aaa aaaa a a a aaaaaaa aa aaaaaa aa aaaaaaaaaaa. Aaaaaaaaa aa aaaaaaaa, aaaaaaaaa aa aaaaaa aaaaaaaaa a aaaaaaa aa a aaaaaaaa aaa aa aaaaa. Aaaaaaa aa aaa aaaa aaaaaaaaaa aaaaa. Aaaaaa aaa a aaaaa aaa aaaaaaaaa aa aaaaaaaaa, aaaaaa aaaaaaa!\n\nAa aa aaaaa aaaaa aaaaa a aaa aaaa, aaa aaaaaaaa aaaaaa a aaaaa aa aaaaaaaaaa a aaaaaaaaaaa.',
            textAlign: pw.TextAlign.justify,
          style: pw.TextStyle(
            height: 1.5,
            fontSize: 14,
          ),
          ),
            pw.SizedBox(height: 20,),
          pw.Text('Aaaaaaa aaa aaaaaa aa Aaaaaaaaaaa, aaaa aa aaa aaa aaa aaaaaaaaaaaaa aaaaaaaaaaa aaaa a aaaaaaaaaaa aa aaaaaaaa a aaaaa aaaaaaaaaaa a aaaaaaaaaa a aaaaaaaaaa a aaaaaaaaaa, aaaaaaaaaaa a aaaaa aa aaaaaaa Aaaaaaaaaaa!\n\nAa aaaa aaaaaaaa aa aaaaaaaa aaa aaaaaa aaaa aaaaa aaaaaaa, aaaa a aaa aaaa, aaaaaaaa a aaaaa aaaaaaa a aaaaaaaa aaaaaa. Aaaa a aaaa aa aaaaaaaaaa aa aaa aaa-aaaaa.',
            textAlign: pw.TextAlign.justify,
            style: pw.TextStyle(
              height: 1.5,
              fontSize: 14,
            ),),
    ],
    )
    );

    String path = await _localPath;

    final File file = File('$path/example.pdf');
    file.writeAsBytesSync(doc.save());
    print('File saved');

    OpenFile.open('$path/example.pdf');
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        // Here we take the value from the MyHomePage object that was created by
        // the App.build method, and use it to set our appbar title.
        title: Text(widget.title),
      ),
      body: Center(
        // Center is a layout widget. It takes a single child and positions it
        // in the middle of the parent.
        child: Column(

          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            Text(
              'You have pushed the button this many times:',
            ),
            Text(
              '$_counter',
              style: Theme.of(context).textTheme.display1,
            ),
          ],
        ),
      ),
      floatingActionButton: FloatingActionButton(
        onPressed: _incrementCounter,
        tooltip: 'Increment',
        child: Icon(Icons.add),
      ), // This trailing comma makes auto-formatting nicer for build methods.
    );
  }
}
```

**Expected behavior**
Expected behavior is to lastline of a paragraph to leftAlign in justify textAlign text with direction ltr and rightAlign in rtl.

**Screenshots**
![image](https://user-images.githubusercontent.com/898113/80409638-95a4bb00-889f-11ea-9d48-3bdac8009ee2.png)

**Flutter Doctor**
Happens in latest version of pdf and flutter

**Desktop (please complete the following information):**
 - [x] iOS
 - [X] Android
 - [ ] Browser

**Smartphone (please complete the following information):**
N/A

**Additional context**
I did some work around to solve my problem, spliting paragraphs breaks in mutiple widgets. I think I can fix it, but since I am not very knowledgable of the plugin I think it is better for owner to fix. I hope I made it easy to replicate the problem and how to solve it.

**Comment by DavBfr** [comment]
Thanks, I will check it. If you want to create a PR, that's event better!

**Comment by DavBfr** [comment]
I think just putting `true` in pdf\lib\text.dart line 647 fixes the issue. Can you confirm and check that everything else is correct?

**Comment by jeanadam** [comment]
Yep, perfectly fixed. No more delta infinity as well.

**Comment by DavBfr** [comment]
Fixed in pdf 1.7.1

**Comment by DavBfr** [comment]
Don't hesitate to [Buy Me A Coffee](https://www.buymeacoffee.com/JORBmbw9h).
